scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala

```diff
@@ -0,0 +1,97 @@
+package scalafix
+package internal.patch
+
+import scala.meta._
+
+object RangePosition {
+  def apply(position: Position): RangePosition =
+    RangePosition(position.start, position.end)
+}
+
+case class RangePosition(start: Int, end: Int) {
+  def withIn(other: RangePosition): Boolean = {
+    start <= other.start && other.end <= end
+  }
+}
+case class PointPosition(offset: Int)
+
+case class CommentFilter(
+    start: PointPosition,
+    end: Option[PointPosition],
+    rules: Set[String]
+) {
+  def isEnabled(ruleName: String, rulePosition: RangePosition): Boolean = {
+    val ruleMatches =
+      if (rules.nonEmpty) rules.contains(ruleName)
+      else true
+
+    ruleMatches && intersect(rulePosition)
+  }
+
+  private def intersect(rulePosition: RangePosition): Boolean = {
+    end match {
+      case Some(endPos) => {
+        val range = RangePosition(start.offset, endPos.offset)
+        range.withIn(rulePosition)
+      }
+      case None => {
+        start.offset <= rulePosition.start
+      }
+    }
+  }
+}
+
+sealed trait FilterAnchor {
+  def position: PointPosition
+}
+
+object FilterAnchor {
+  case class Start(position: PointPosition, rules: Set[String])
+      extends FilterAnchor
+  case class End(position: PointPosition) extends FilterAnchor
+}
+
+object EscapeHatch {
+  private val FilterAnchorStart = "scalafix:off (.*)".r
+  private val FilterAnchorEnd = "scalafix:on".r
+
+  def apply(tree: Tree): EscapeHatch = {
+    val anchors: List[FilterAnchor] =
+      tree.tokens.collect {
+        case comment @ Token.Comment(FilterAnchorStart(rules)) => {
+          FilterAnchor.Start(
+            PointPosition(comment.pos.end),
+            rules.split("\\s+").toSet
+          )
+        }
+
+        case comment @ Token.Comment(FilterAnchorEnd()) => {
+          FilterAnchor.End(
+            PointPosition(comment.pos.start)
+          )
+        }
+      }.toList
+
+    val filters: List[CommentFilter] =
+      anchors.sortBy(_.position.offset).grouped(2).toList.collect {
+        case List(FilterAnchor.Start(start, rules), FilterAnchor.End(end)) => {
+          CommentFilter(start, Some(end), rules)
+        }
+
+        case List(FilterAnchor.Start(start, rules)) => {
+          CommentFilter(start, None, rules)
+        }
+
+        case _ => sys.error("invalid anchors")
+      }
+
+    new EscapeHatch(filters)
+  }
+}
+
+class EscapeHatch(filters: List[CommentFilter]) {
```

Comment:
I feel like we can simplify the logic a bit if we change the semantics to be

> a rule `r` is disabled in position `p` if there is a comment disabling `r` at position `p1 < p` and there is no comment enabling `r` in position `p2` where `p1 < p2 < p`.

This does not handle push/pop semantics like
```
// scalafix:off A
// scalafix:off A
// scalafix:on A
object ReportA // A is reported because there is a single `on:A` between it and `off:A`
```
But I question how useful push/pop is in cases like that, `// scalafix:on A` should IMO always enable rule `A` no matter how many leading `off: A` there are in the file.


```scala
class EscapeHatch(
    starts: TreeMap[Int, FilterMatcher],
    ends: TreeMap[Int, FilterMatcher]) {
  def isEnabled(id: String, pos: Int): Boolean = !isDisabled(id, pos)
  def isDisabled(id: String, pos: Int): Boolean = {
    starts.to(pos).exists {
      case (spos, sname) =>
        sname.matches(id) && {
          !ends.range(spos, pos + 1).exists {
            case (epos, ename) =>
              ename.matches(id)
          }
        }
    }
  }
}
object EscapeHatch {
  case class Comment(pos: Int, id: String, kind: Kind)
  sealed abstract class Kind
  case object On extends Kind
  case object Off extends Kind
  def apply(comments: List[Comment]): EscapeHatch = {
    val starts = TreeMap.newBuilder[Int, FilterMatcher]
    val ends = TreeMap.newBuilder[Int, FilterMatcher]
    comments.foreach {
      case Comment(pos, id, Off) => starts += (pos -> FilterMatcher(id))
      case Comment(pos, id, On) => ends += (pos -> FilterMatcher(id))
    }
    new EscapeHatch(starts.result(), ends.result())
  }
}
```
`FilterMatcher` comes from internal.config. One benefit of this approach is that we avoid having to match on/off comments.

What do you think?

Reply:
I think it's easier te reason about scalafix:on|off as brackets `{` and `}` so if I convert the thing above in bracket notation it look like:

```
{
{
}
```

Reply:
I asked for a third opinion (Martin D.) and he said the way you propose makes more sense. I agree with you guys now ;-).